test/hotspot/jtreg/gc/TestAlwaysPreTouchBehavior.java

```diff
@@ -28,103 +28,153 @@
+ * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xmx64m gc.TestAlwaysPreTouchBehavior -XX:+UnlockExperimentalVMOptions -XX:+UseEpsilonGC
  */
 
 
 import jdk.test.lib.Asserts;
 
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
 import jdk.test.whitebox.WhiteBox;
+import jtreg.SkippedException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 public class TestAlwaysPreTouchBehavior {
 
-    public static void main(String [] args) {
-    long rss = WhiteBox.getWhiteBox().rss();
-    System.out.println("RSS: " + rss);
-    if (rss == 0) {
-        System.out.println("cannot get RSS, just skip");
-        return; // Did not get available RSS, just ignore this test.
+    final static WhiteBox wb = WhiteBox.getWhiteBox();
+
+    final static long M = 1024 * 1024;
+    final static long heapsize = M * 128;
+    final static long expectedMaxNonHeapRSS = M * 256;
+    final static  long requiredAvailableBefore = heapsize * 2 + expectedMaxNonHeapRSS;
+    final static  long requiredAvailableDuring = expectedMaxNonHeapRSS;
+
+    private static String[] prepareOptions(String[] extraVMOptions) {
+        List<String> allOptions = new ArrayList<String>();
+        if (extraVMOptions != null) {
+            allOptions.addAll(Arrays.asList(extraVMOptions));
+        }
+        allOptions.add("-Xmx" + heapsize);
+        allOptions.add("-Xms" + heapsize);
+        allOptions.add("-XX:+AlwaysPreTouch"); // Stabilize RSS
+        allOptions.add("-XX:+UnlockDiagnosticVMOptions"); // For whitebox
+        allOptions.add("-XX:+WhiteBoxAPI");
+        allOptions.add("-Xbootclasspath/a:.");
+        allOptions.add("-XX:-ExplicitGCInvokesConcurrent"); // Invoke explicit GC on System.gc
+        allOptions.add("--add-exports=java.base/jdk.internal.misc=ALL-UNNAMED");
+        allOptions.add(TestAlwaysPreTouchBehavior.class.getName());
+        allOptions.add("run");
+        return allOptions.toArray(new String[0]);
+    }
+
+    private static OutputAnalyzer runTestWithOptions(String[] extraOptions) throws IOException {
+        ProcessBuilder pb = ProcessTools.createLimitedTestJavaProcessBuilder(prepareOptions(extraOptions));
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldHaveExitValue(0);
+        output.reportDiagnosticSummary();
+        return output;
+    }
+
+    public static void main(String [] args) throws IOException {
+
+        Runtime runtime = Runtime.getRuntime();
+        long committed = runtime.totalMemory();
+        long avail = wb.hostAvailableMemory();
+        long rss = wb.rss();
+        System.out.println("RSS: " + rss + " available: " + avail + " committed " + committed);
+
+        if (args[0].equals("run")) {
```

Comment:
When will this branch be taken? I can't find where the `run` arg is specified.

Reply:
See line 141

Reply:
I see; thanks. Can you add a comment referencing `prepareOptions`?

Reply:
Sure.